### PR TITLE
Issue 12 search for test classes

### LINKED ISSUE #12: Package in-addition to of Individual test class names
As a user, I want the option to use atam4j by simply passing a package name containing my test classes instead of passing individual test class names so that adding new tests to the suite is easier and seamless and does not need modification to the test runner each time.

### PR DESCRIPTION
**Pull request by tjcutajar**
Instead of passing in an array of test classes to run, search for classes annotated with @Monitor (happy to change the name). I've kept the existing constructors that take a Class[] and made them package private for testing purposes. Closes #12.

### CODE REVIEW

**Comment by phillbarber** [comment]
Hi Tyrone,
   Thanks so much for contributing to atam4j, much appreciated.

Introducing the @Monitor annotation is an interesting idea.  On the practical side of things I think it's definitely an improvement from specifying a list of classes.  Conceptually however, I'm not sure that we should add a dependency from the tests to atam4j.  I'd rather atam4j could just run plain old junit tests that didn't need any modification whatsoever.

That said, maybe I'm a biased anti annotations man!  Perhaps we could add support for specifying tests to run via direct packages/class reference or Annotation mode?

Thoughts?  Anurag... Thoughts?

Also - just to make you aware - I'm still waiting for another pull request to be merged by Anurag (see #23).  Feel free to shoot with nerf guns of power proportionate to frustration.

**Comment by TJ-Cutajar** [comment]
I see your point. The motivation for using annotations here is firstly that it seems slightly odd to pass in package names to a library and secondly what if you don't want all of the tests in the specified package to be monitored. For sure we could provide options here, namely using the annotation, specifying a package name or providing a list of classes.

I'll rework this PR once(if) #23  goes in :) @anuragkapur.

**Comment by anuragkapur** [comment]
Nice one. I agree with the conclusion you guys have already come to. Class names array, package name or annotation - all may be beneficial in different use-cases - so lets have them as options that the client can pick from!

**Comment by phillbarber** [comment]
Nice one!  Supporting both options sounds great to me.  Thanks all!

On Mon, May 18, 2015 at 9:58 AM, Anurag Kapur <notifications@github.com>
wrote:

> Nice one. I agree with the conclusion you guys have already come to. Class
> names array, package name or annotation - all may be beneficial in
> different use-cases - so lets have them as options that the client can pick
> from!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/atam4j/atam4j/pull/25#issuecomment-102980957>.
>



-- 

Pergola Software Limited

Company number:  09275615 (incorporated in England and Wales)
Registered office:  5 Beaumont Gate, Shenley Hill,  Radlett,  WD7 7AR

**Comment by TJ-Cutajar** [comment]
@phillbarber @anuragkapur I've rebased with latest master. We now have the option to build the AcceptanceTestHealthCheckManager with either the 'withTestClasses' or 'withMonitorAnnotationScanningEnabled' option.

File: README.md [hunk 1]
```diff
@@ -23,7 +23,7 @@ Acceptance Tests As Monitors
 
 1. Include the atam4j maven dependency.
 
-2. Write Junit based tests in the usual manner, with the exception of including them in the main application classpath 
+2. Write Junit based tests in the usual manner, with the exception of adding a `@Monitor` annotation to the class and including them in the main application classpath
```
Comment (anuragkapur) [inline README.md]: @TJ-Cutajar - Given the annotation isn't mandatory, we should update the sentence?

Comment (TJ-Cutajar) [inline README.md]: Done.

**Comment by TJ-Cutajar** [comment]
Updated the readme. Closes #27

**Comment by phillbarber** [comment]
Hi Tyrone,
   Looks good, however I think that the code which does the reflection to find the @Monitor instances should be in the builder class.  The AcceptanceTestsRunnerTask can then just take an array of Classes and not an optional and not be concerned with finding its dependencies.    

Cheers again

**Comment by tjcutajar** [comment]
@phillbarber @anuragkapur I've rebase with latest master and moved the annotation scanning to the builder.

File: README.md [hunk 2]
```diff
@@ -25,13 +25,13 @@ Acceptance Tests As Monitors
...
+`src/test/java` directory. You can choose to either add a `@Monitor` annotation to your test classes and let atam4j detect them or 
+simply supply an array of classes to the builder.
 
-3. Instantiate  `Atam4j` in the `run` method of your dropwizard application class.    
+3. Instantiate `Atam4j` in the `run` method of your dropwizard application class.    
 
-        new Atam4j.Atam4jBuilder()     
-            .withTestClasses(new Class[]{HelloWorldTest.class})     
-            .withEnvironment(environment)      
+        new Atam4j.Atam4jBuilder(environment)
```
Comment (anuragkapur) [inline README.md]: `withEnvironment(environment)` is mandatory and should be included in the sample snippet. Am i missing something?

**Comment by anuragkapur** [comment]
Though I had a question/comment, I have gone ahead and merged :8ball: this as it wasn't critical. I will update the readme as a separate after @tjcutajar has had the chance to review and respond to my comment above...

Thanks @tjcutajar for the changes :+1:

File: README.md (hunk 2 above)
Comment (anuragkapur) [inline README.md]: Ahh, sorry, I see you have moved that to the constructor. Ignore my comment :)

**Pull request merged**